src/equisolve/nn/module_tensor.py

```diff
@@ -0,0 +1,172 @@
+try:
+    from metatensor.torch import Labels, LabelsEntry, TensorBlock, TensorMap
+
+    HAS_METATENSOR_TORCH = True
+except ImportError:
+    from metatensor import Labels, TensorBlock, TensorMap
+    from metatensor.core.labels import LabelsEntry
+
+    HAS_METATENSOR_TORCH = False
+
+from copy import deepcopy
+from typing import List, Optional
+
+import torch
+from torch.nn import Linear, Module, ModuleDict
+
+
+@torch.jit.interface
+class ModuleTensorMapInterface(torch.nn.Module):
+    """
+    This interface required for TorchScript to index the ModuleDict with non-literals
+    in ModuleTensorMap. Any module that is used with ModuleTensorMap must implement
+    this interface to be TorchScript compilable.
+
+    Note that the *typings and argument names must match exactly* so that an interface
+    is correctly implemented.
+
+    Reference
+    ---------
+    https://github.com/pytorch/pytorch/pull/45716
+    """
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        pass
+
+
+class ModuleTensorMap(Module):
+    """
+    ModuleDictTensorMap is needed when the keys correspond to different features sizes
+
+    :param module_map: a dictionary of modules with tensor map keys as dict keys
+        each module is applied on a block
+
+    :param out_tensor: a tensor map that has the output properties of the Labels
+    """
+
+    def __init__(self, module_map: ModuleDict, out_tensor: Optional[TensorMap] = None):
```

Comment:
I am not sure what the `out_tensor` is doing. Is this the `TensorMap` where the results will be written in?